### PR TITLE
Integrate Apple changes to mechanism glue

### PR DESCRIPTION
**Pull request by lhoward**
WIP integration of Heimdal-520 mechglue from Apple on top of lukeh/gss-cred-store branch

### CODE REVIEW

File: lib/gssapi/gssapi/gssapi_plugin.h [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+ * SUCH DAMAGE.
+ */
+
+#ifndef __GSSAPI_PLUGIN_H
+#define __GSSAPI_PLUGIN_H 1
+
+#define GSSAPI_PLUGIN "gssapi_plugin"
+
+typedef gss_cred_id_t
+(*gssapi_plugin_isc_replace_cred)(gss_const_name_t target, gss_OID mech, gss_const_cred_id_t original_cred, OM_uint32 flags);
```
Comment (nicowilliams) [inline lib/gssapi/gssapi/gssapi_plugin.h]: I strongly object to any GSS functions, API or SPI, without a `OM_uint32 *minor_status` argument.  See my `pgss` branch to understand why.

**Comment by nicowilliams** [comment]
This will take a long time to review...  I've set the milestone to Heimdal 9.  I will make this a low-priority for now.

**Comment by lhoward** [comment]
Here's a summary of changes vs lukeh/gss-cred-store (once that's integrated, the changes will be smaller). Maybe we could pick up the bugfixes for that branch.

* krb5 plugin API for replacing *GSS* credentials
* move `gss_auth_identity_desc` into new header, gssapi_spi.h
* mechglue logging API `_gss_mg_log()`
* `_gss_mg_get_error()/_gss_mg_set_error()` drop major status, handle `GSS_C_MECH_CODE` only
* new helpers for allocating creds and names
* `GM_USE_MG_CRED` flag (not set by any mechs yet) for using mechglue creds for pseudo-mechs; note, if I read the code right, this means that you acquire creds for the concrete mechs and pass to SPNEGO (which breaks tests that only acquire creds for SPNEGO)
* new helpers APIs that are used by the mechs themselves, e.g. `gss_mg_gen_cb()`
* new top-level APIs for existing SPI such as `gss_cred_hold()`

Bugfixes:

* check `ctx->gc_ctx` in `gss_inquire_context()`
* check `gm_inquire_cred != NULL` in `gss_inquire_context()`
* check `min_lifetime` in `gss_inquire_context()`
* check `gm_inquire_cred_by_mech != NULL` in `gss_inquire_cred_by_mech()`
* set mech error in `gss_inquire_cred_by_mech()`
* don't clobber error in `gss_inquire_cred_by_mech()`
* don't pass NULL minor_status to `gss_krb5_free_lucid_sec_context()`
* allow NULL ccache in `gss_krb5_ccache_name()`

Things I didn't merge:

* The mechanism changes themselves
* Automatically generated prototypes
* Exporting some of the new API

**Comment by nicowilliams** [comment]
Thanks for that summary.  It really helps.

**Comment by lhoward** [comment]
Now if you look at the differences between lukeh/gss-cred-store and lukeh/apple-gss-mechglue, you'll find they're not very many. Sneaky! ;)

**Comment by lhoward** [comment]
Closing, it's not going to happen.

**Pull request closed without merging**